doc/tutorial/concepts/index.md

```diff
@@ -0,0 +1,311 @@
+  * The (rather dense) [documentation on cppreference](https://en.cppreference.com/w/cpp/language/constraints).
+
+# Constraints
+
+## Motivation
+
+Generic programming is all about making your algorithms work with different types, e.g. the local alignment algorithm
+itself is (almost) the same for nucleotide- and aminoacid-sequences so you should be able use the same code
+to model your solution.
+
+In C++ this is achieved through the use of templates.
+A template is a placeholder for a type.
+As soon as you use the template with a concrete type, an actual implementation
+of the function/class/variable-template will be *instantiated* by the compiler, i.e. the compiler generates two
+copies of your algorithm, one for the sequences of seqan3::dna5, and one for sequences of type seqan3::aa27.
+
+A typical template parameter is said to be *unconstrained*, i.e. it can be instantiated with any type:
+
+```cpp
+template <typename t>
+t add(t const v1, t const v2)
+{
+    return v1 + v2;
+}
+
+int main()
+{
+    return add(1, 3); // instantiates add<int>()
+}
+```
+
+Of course this will not actually compile for all types, because the function template **implicitly requires** that
+types provide a `+` operator.
+If a type is used that does not have a `+` operator, this implicitness causes the compiler to fail at the place where
+such operator is used – and not at the place the template is instantiated.
+This leads to very complex error messages for deeply nested code.
+
+*Constraints* are a way of making requirements of template arguments **explicit**.
+Constraints can be formulated ad-hoc, but this tutorial only covers *concepts*.
+The interested reader can check the [documentation](https://en.cppreference.com/w/cpp/language/constraints) to learn
+about ad-hoc definitions.
+Concepts are a set of constraints with a given name.
+Let's assume there is a concept called `Addable` that requires the existance of a `+` operator (as previously mentioned
+the syntax for defining concepts is not covered here).
+The following snippet demonstrates how we can constrain our function template, i.e. make the template immediately
+reject any types that don't satisfy the requirement:
+
+```cpp
+template <Addable t>
+t add(t const v1, t const v2)
+{
+    return v1 + v2;
+}
+
+int main()
+{
+    return add(1, 3); // instantiates add<int>()
+}
+```
+
+The only difference is that we have replaced `typename` with `Addable`.
+If you plug in a type that does not model `Addable` you will get a message stating exactly that and not a cryptic
+template trace.
+
+The standard library provides a set of [predefined concepts](https://en.cppreference.com/w/cpp/concepts).
+For our example above, the std::Integral concept could have been used.
+
+## Syntax variants
+
+Depending on the complexity of your constraint statements, three different syntaxes are available to enforce
+constraints; all of the following are equivalent.
```

Comment:
```suggestion
constraints; all of the following are equivalent:
```

Reply:
but there are colons again. double colon is weird